src/source-updater.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * source-updater
+ *
+ * A queue of callbacks to be serialized and applied when a
+ * MediaSource and its associated SourceBuffers are not in the
+ * updating state. It is used by the segment loader to update the
+ * underlying SourceBuffers when new data is loaded, for instance.
+ *
+ */
+(function(window, videojs) {
+  'use strict';
+
+  /**
+   * Construct a new SourceUpdater using the specified MediaSource.
+   * @param mediaSource {MediaSource} the MediaSource to create the
+   * SourceBuffer from
+   * @param mimeType {string} the desired MIME type of the underlying
+   * SourceBuffer
+   */
+  videojs.Hls.SourceUpdater = videojs.extend(null, {
+    constructor: function(mediaSource, mimeType) {
+      var createSourceBuffer = function() {
+        this.sourceBuffer_ = mediaSource.addSourceBuffer(mimeType);
+
+        // run completion handlers and process callbacks as updateend
+        // events fire
+        this.sourceBuffer_.addEventListener('updateend', function() {
+          var pendingCallback = this.pendingCallback_;
+          this.pendingCallback_ = null;
+
+          if (pendingCallback) {
+            pendingCallback();
+          }
+        }.bind(this));
+        this.sourceBuffer_.addEventListener('updateend',
+                                            this.runCallback_.bind(this));
+
+        this.runCallback_();
+      }.bind(this);
+
+      this.callbacks_ = [];
+      this.pendingCallback_ = null;
+      this.timestampOffset_ = 0;
+
+      if (mediaSource.readyState === 'closed') {
+        mediaSource.addEventListener('sourceopen', createSourceBuffer);
+      } else {
+        createSourceBuffer();
+      }
+    },
+
+    /**
+     * Queue an update to append an ArrayBuffer.
+     * @see http://www.w3.org/TR/media-source/#widl-SourceBuffer-appendBuffer-void-ArrayBuffer-data
+     */
+    appendBuffer: function(bytes, done) {
+      this.queueCallback_(function() {
+        this.sourceBuffer_.appendBuffer(bytes);
+      }, done);
+    },
+
+    /**
+     * Indicates what TimeRanges are buffered in the managed SourceBuffer.
+     * @see http://www.w3.org/TR/media-source/#widl-SourceBuffer-buffered
+     */
+    buffered: function() {
+      if (!this.sourceBuffer_) {
+        return videojs.createTimeRanges();
+      }
+      return this.sourceBuffer_.buffered;
+    },
+
+    /**
+     * Queue an update to set the duration.
+     * @see http://www.w3.org/TR/media-source/#widl-MediaSource-duration
+     */
+    duration: function(duration) {
```

Comment:
Note for the future: this doesn't work at all. Duration is set in MSE on the MediaSource, not the SourceBuffer, and requires all attached source buffers to be in an updatable state. That might mean we need an abstraction that manages updates all the way up at the MediaSource level but that seems complicated so I've skipped it so far.